toolchain-clojure/src/main/java/com/sourcegraph/toolchain/clojure/ClojureParseTreeListener.java

```diff
@@ -0,0 +1,417 @@
+import com.sourcegraph.toolchain.core.objects.DefKey;
+import com.sourcegraph.toolchain.core.objects.Ref;
+import com.sourcegraph.toolchain.language.Scope;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+
+import java.util.StringJoiner;
+
+
+class ClojureParseTreeListener extends ClojureBaseListener {
+
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClojureParseTreeListener.class);
+
+    public static final char PATH_SEPARATOR = '.';
+
+    private LanguageImpl support;
+
+//    private NamespaceContextResolver nsContextResolver = new NamespaceContextResolver.getInstance();
+
+    private NamespaceContextResolver nsContextResolver = new NamespaceContextResolver();
+
+    //Context<Boolean> context = new Context<>();
+
+    private Map<ParserRuleContext, Boolean> defs = new IdentityHashMap<>();
+
+    ClojureParseTreeListener(LanguageImpl support) {
+        this.support = support;
+    }
+
+    @Override
+    public void enterFn_binding(ClojureParser.Fn_bindingContext ctx) {
+        nsContextResolver.context().enterScope(nsContextResolver.context().currentScope().next(PATH_SEPARATOR));
+        List<ClojureParser.ParameterContext> params = ctx.arguments().parameter();
+        saveParametersInScope(params);
+    }
+
+    @Override
+    public void exitFn_binding(ClojureParser.Fn_bindingContext ctx) {
+        nsContextResolver.context().exitScope();
+    }
+
+    @Override
+    public void enterAnonym_fn_binding(ClojureParser.Anonym_fn_bindingContext ctx) {
+        nsContextResolver.context().enterScope(nsContextResolver.context().currentScope().next(PATH_SEPARATOR));
+        List<ClojureParser.ParameterContext> params = ctx.arguments().parameter();
+        saveParametersInScope(params);
+    }
+
+    @Override
+    public void exitAnonym_fn_binding(ClojureParser.Anonym_fn_bindingContext ctx) {
+        nsContextResolver.context().exitScope();
+    }
+
+    @Override
+    public void enterLetfn_function_binding(ClojureParser.Letfn_function_bindingContext ctx) {
+        nsContextResolver.context().enterScope(nsContextResolver.context().currentScope().next(PATH_SEPARATOR));
+        List<ClojureParser.ParameterContext> params = ctx.arguments().parameter();
+        saveParametersInScope(params);
+    }
+
+    @Override
+    public void exitLetfn_function_binding(ClojureParser.Letfn_function_bindingContext ctx) {
+        nsContextResolver.context().exitScope();
+    }
+
+    @Override
+    public void enterSimple_fn_def(ClojureParser.Simple_fn_defContext ctx) {
+        ClojureParser.Fn_nameContext nameCtx = ctx.fn_name();
+        String fnStartKeyword = ctx.fn_start().getText();
+
+        enterFunctionWithName(nameCtx, fnStartKeyword, formParamsDoc(ctx.arguments().parameter(), ctx.last_arguments()));
```

Comment:
Please cache the results of `node::somechildren()` because each call traverses all the children nodes